apps/console/src/features/roles/components/edit-role/edit-role-common/permissions-list.tsx

```diff
@@ -0,0 +1,231 @@
+import { IdentifiableComponentInterface } from "@wso2is/core/models";
+import React, { FunctionComponent, HTMLAttributes, ReactElement, SyntheticEvent, useEffect, useState } from "react";
+import { useTranslation } from "react-i18next";
+import { AutoCompleteRenderOption } from "./auto-complete-render-option";
+import { RenderChip } from "./render-chip";
+import { APIResourceInterface, ScopeInterface } from "../../../models/apiResources";
+
+interface PermissionsListPropsInterface extends  IdentifiableComponentInterface {
+    /**
+     * API resource.
+     */
+    apiResource: APIResourceInterface;
+    /**
+     * Initial selected permissions.
+     */
+    initialSelectedPermissions?: ScopeInterface[];
+    /**
+     * Selected permissions.
+     */
+    selectedPermissions: ScopeInterface[];
+    /**
+     * Callback to handle API resource removal.
+     */
+    onChangeScopes: (apiResource: APIResourceInterface, scopes: ScopeInterface[]) => void;
+}
+
+export const PermissionsList: FunctionComponent<PermissionsListPropsInterface> = 
+    (props: PermissionsListPropsInterface): ReactElement => {
+
+        const {
+            apiResource,
+            initialSelectedPermissions,
+            selectedPermissions,
+            onChangeScopes,
+            [ "data-componentid" ]: componentId
+        } = props;
+
+        const { t } = useTranslation();
+
+        const [ activeOption, setActiveOption ] = useState<ScopeInterface>(undefined);
+        const [ removedPermissions, setRemovedPermissions ] = useState<ScopeInterface[]>([]);
+
+        /**
+         * Set removed permissions.
+         */
+        useEffect(() => {
+            if (initialSelectedPermissions && selectedPermissions) {
+                setRemovedPermissions(initialSelectedPermissions?.filter((permission: ScopeInterface) => {
+                    return selectedPermissions?.find(
+                        (selectedPermission: ScopeInterface) =>
+                            selectedPermission.name === permission.name) === undefined;
+                }));
+            }
+        }, [ initialSelectedPermissions, selectedPermissions ]);
+
+        /**
+         * Handles the select scope action.
+         */
+        const handleScopeSelection = (scopes: ScopeInterface[]) => {
+            onChangeScopes(apiResource, scopes);
+        };
+
+        /**
+         * Handle the restore permissions.
+         * 
+         * @param remainingPermissions - remaining permissions
+         */
+        const handleRestorePermissions = (remainingPermissions: ScopeInterface[]) => {
+            const removedRoles: ScopeInterface[] = [];
+
+            removedPermissions.forEach((permission: ScopeInterface) => {
+                if (!remainingPermissions?.find((remainingPermission: ScopeInterface) =>
+                    remainingPermission.name === permission.name)) {
+                    removedRoles.push(permission);
+                }
+            });
```

Comment:
We can make the logic more expressive and concise by using filter and some functions.
`some` is more fitting for knowing if there's any match (boolean).
`filter` --> `forEach` + `find`
```
const removedRoles: ScopeInterface[] = removedPermissions.filter(permission => 
    !remainingPermissions?.some(remainingPermission => remainingPermission.name === permission.name)
);
```
WDYT?

Reply:
Fixed with ed31181181da4228a981a4e356b5f55ac42ce0db